### DISCUSSION TITLE
Need a tutorial for pretraining

### DISCUSSION THREAD

**Discussion by RaphaelDuan** [speechbrain/speechbrain#1806] [Q&A]
### Describe the bug

I'm trying to use the WSJ02mix pre-trained model to fine-tune my custom dataset instead of training from scratch. However, I can not find a tutorial on how to do this. The code is hard to understand and confused me a lot. I have two questions:

1. I think it can load the pre-trained model from the hugging face and fine-tune the model using a custom dataset. How to do it? Could anyone show me **an example code**?
2. Since I want to modify some structure of the Sepformer network and will cause the size of some layers' parameters not to be matchable with the pre-trained model. What if I **only want to load part of the pre-trained model** (e.g., some layers, some blocks), what should I do?

I tried to use the Pretrainer class as the below codes, but it does not work. Please help me.

### Expected behaviour

**Example code** of using WSJ0 pre-trained model to fine-tune my custom dataset by only load **part of** the pre-trained model.

### To Reproduce

**Python code:**

      # Brain class initialization
      separator = Separation(
          modules=hparams["modules"],
          opt_class=hparams["optimizer"],
          hparams=hparams,
          run_opts=run_opts,
          checkpointer=hparams["checkpointer"],
      )
  
      # Load pretrained model if pretrained_separator is present in the yaml
      if "pretrained_separator" in hparams:
          # Load pretrained model
          from speechbrain.utils.parameter_transfer import Pretrainer
  
          # Initialization of the pre-trainer
          pretrain = Pretrainer(loadables={'model': separator},
                                paths={'model': 'speechbrain/sepformer-wsj02mix/masknet.ckpt'})
  
          # We download the pretrained model from HuggingFace in this case
          pretrain.collect_files()
          pretrain.load_collected(device='cuda:0')
  
      # re-initialize the parameters if we don't use a pretrained model
      if "pretrained_separator" not in hparams:
          for module in separator.modules.values():
              separator.reset_layer_recursively(module)

**My yaml:**
    
    seed: 1234
    __set_seed: !apply:torch.manual_seed [!ref <seed>]
    
    # Data params
    
    # e.g. '/yourpath/wsj0-mix/2speakers'
    # end with 2speakers for wsj0-2mix or 3speakers for wsj0-3mix
    data_folder: !PLACEHOLDER
    
    # the path for wsj0/si_tr_s/ folder -- only needed if dynamic mixing is used
    # e.g. /yourpath/wsj0-processed/si_tr_s/
    # you need to convert the original wsj0 to 8k
    # you can do this conversion with the script ../meta/preprocess_dynamic_mixing.py
    base_folder_dm: /yourpath/wsj0-processed/si_tr_s/
    pretrained_separator: speechbrain/sepformer-wsj02mix
    
    experiment_name: sepformer
    output_folder: !ref results/<experiment_name>/<seed>
    train_log: !ref <output_folder>/train_log.txt
    save_folder: !ref <output_folder>/save
    train_data: !ref <save_folder>/wsj_tr.csv
    valid_data: !ref <save_folder>/wsj_cv.csv
    test_data: !ref <save_folder>/wsj_tt.csv
    skip_prep: False
    
    
    # Experiment params
    auto_mix_prec: True # Set it to True for mixed precision
    test_only: False
    num_spks: 2 # set to 3 for wsj0-3mix
    noprogressbar: False
    save_audio: True # Save estimated sources on disk
    n_audio_to_save: 20
    sample_rate: 8000
    
    # Training parameters
    N_epochs: 200
    batch_size: 1
    lr: 0.00015
    clip_grad_norm: 5
    loss_upper_lim: 999999  # this is the upper limit for an acceptable loss
    # if True, the training sequences are cut to a specified length
    limit_training_signal_len: True
    # this is the length of sequences if we choose to limit
    # the signal length of training sequences
    training_signal_len: 80000
    
    # Set it to True to dynamically create mixtures at training time
    dynamic_mixing: False
    
    # Parameters for data augmentation
    use_wavedrop: False
    use_speedperturb: True
    use_rand_shift: False
    min_shift: -8000
    max_shift: 8000
    
    speedperturb: !new:speechbrain.lobes.augment.TimeDomainSpecAugment
        perturb_prob: 1.0
        drop_freq_prob: 0.0
        drop_chunk_prob: 0.0
        sample_rate: !ref <sample_rate>
        speeds: [95, 100, 105]
    
    wavedrop: !new:speechbrain.lobes.augment.TimeDomainSpecAugment
        perturb_prob: 0.0
        drop_freq_prob: 1.0
        drop_chunk_prob: 1.0
        sample_rate: !ref <sample_rate>
    
    # loss thresholding -- this thresholds the training loss
    threshold_byloss: True
    threshold: -30
    
    # Encoder parameters
    N_encoder_out: 256
    out_channels: 256
    kernel_size: 16
    kernel_stride: 8
    
    # Dataloader options
    # Set num_workers: 0 on MacOS due to behavior of the multiprocessing library
    dataloader_opts:
        batch_size: !ref <batch_size>
        num_workers: 0
    
    
    # Specifying the network
    Encoder: !new:speechbrain.lobes.models.dual_path.Encoder
        kernel_size: !ref <kernel_size>
        out_channels: !ref <N_encoder_out>
    
    
    SBtfintra: !new:speechbrain.lobes.models.dual_path.SBTransformerBlock
        num_layers: 8
        d_model: !ref <out_channels>
        nhead: 8
        d_ffn: 1024
        dropout: 0
        use_positional_encoding: True
        norm_before: True
    
    SBtfinter: !new:speechbrain.lobes.models.dual_path.SBTransformerBlock
        num_layers: 8
        d_model: !ref <out_channels>
        nhead: 8
        d_ffn: 1024
        dropout: 0
        use_positional_encoding: True
        norm_before: True
    
    MaskNet: !new:speechbrain.lobes.models.dual_path.Dual_Path_Model
        num_spks: !ref <num_spks>
        in_channels: !ref <N_encoder_out>
        out_channels: !ref <out_channels>
        num_layers: 2
        K: 250
        intra_model: !ref <SBtfintra>
        inter_model: !ref <SBtfinter>
        norm: ln
        linear_layer_after_inter_intra: False
        skip_around_intra: True
    
    Decoder: !new:speechbrain.lobes.models.dual_path.Decoder
        in_channels: !ref <N_encoder_out>
        out_channels: 1
        kernel_size: !ref <kernel_size>
        stride: !ref <kernel_stride>
        bias: False
    
    optimizer: !name:torch.optim.Adam
        lr: !ref <lr>
        weight_decay: 0
    
    loss: !name:speechbrain.nnet.losses.get_si_snr_with_pitwrapper
    
    lr_scheduler: !new:speechbrain.nnet.schedulers.ReduceLROnPlateau
        factor: 0.5
        patience: 2
        dont_halve_until_epoch: 85
    
    epoch_counter: !new:speechbrain.utils.epoch_loop.EpochCounter
        limit: !ref <N_epochs>
    
    modules:
        encoder: !ref <Encoder>
        decoder: !ref <Decoder>
        masknet: !ref <MaskNet>
    
    checkpointer: !new:speechbrain.utils.checkpoints.Checkpointer
        checkpoints_dir: !ref <save_folder>
        recoverables:
            encoder: !ref <Encoder>
            decoder: !ref <Decoder>
            masknet: !ref <MaskNet>
            counter: !ref <epoch_counter>
            lr_scheduler: !ref <lr_scheduler>
    
    train_logger: !new:speechbrain.utils.train_logger.FileTrainLogger
        save_file: !ref <train_log>

### Relevant log output

```shell
hparams/sepformer.yaml {'debug': False, 'debug_batches': 2, 'debug_epochs': 2, 'device': 'cuda:0', 'data_parallel_backend': False, 'distributed_launch': False, 'distributed_backend': 'nccl', 'find_unused_parameters': False} data_folder: /home/duandi/Dataset/WSJ0/wsj0_mixed/2speakers
pretrained_separator: True
speechbrain.core - Beginning experiment!
speechbrain.core - Experiment folder: results/sepformer/1234
speechbrain.core - Info: auto_mix_prec arg from hparam file is used
speechbrain.core - Info: noprogressbar arg from hparam file is used
speechbrain.core - 25.7M trainable parameters in Separation
speechbrain.pretrained.fetching - Fetch masknet.ckpt: Using existing file/symlink in model_checkpoints/model.ckpt.
speechbrain.utils.parameter_transfer - Loading pretrained files for: model
Traceback (most recent call last):
  File "/home/duandi/Projects/speechbrain-develop/speechbrain/core.py", line 1356, in _recover
    save_dict = yaml.safe_load(f)
  File "/home/duandi/Downloads/ENTER/envs/EarSE/lib/python3.7/site-packages/yaml/__init__.py", line 125, in safe_load
    return load(stream, SafeLoader)
  File "/home/duandi/Downloads/ENTER/envs/EarSE/lib/python3.7/site-packages/yaml/__init__.py", line 79, in load
    loader = Loader(stream)
  File "/home/duandi/Downloads/ENTER/envs/EarSE/lib/python3.7/site-packages/yaml/loader.py", line 34, in __init__
    Reader.__init__(self, stream)
  File "/home/duandi/Downloads/ENTER/envs/EarSE/lib/python3.7/site-packages/yaml/reader.py", line 85, in __init__
    self.determine_encoding()
  File "/home/duandi/Downloads/ENTER/envs/EarSE/lib/python3.7/site-packages/yaml/reader.py", line 124, in determine_encoding
    self.update_raw()
  File "/home/duandi/Downloads/ENTER/envs/EarSE/lib/python3.7/site-packages/yaml/reader.py", line 178, in update_raw
    data = self.stream.read(size)
  File "/home/duandi/Downloads/ENTER/envs/EarSE/lib/python3.7/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 64: invalid start byte
```

### Additional context

The original part in speechbrain/recipes/WSJ0Mix/separation/train.py is like this.

        # Load pretrained model if pretrained_separator is present in the yaml
        if "pretrained_separator" in hparams:
            run_on_main(hparams["pretrained_separator"].collect_files)
            hparams["pretrained_separator"].load_collected()
    
        # Brain class initialization
        separator = Separation(
            modules=hparams["modules"],
            opt_class=hparams["optimizer"],
            hparams=hparams,
            run_opts=run_opts,
            checkpointer=hparams["checkpointer"],
        )
    
        # re-initialize the parameters if we don't use a pretrained model
        if "pretrained_separator" not in hparams:
            for module in separator.modules.values():
                separator.reset_layer_recursively(module)

However, I cannot find any "pretrained_separator" in the default speechbrain/recipes/WSJ0Mix/separation/hparams/sepformer.yaml
Therefore, I made the change like **pretrained_separator: speechbrain/sepformer-wsj02mix** or **pretrained_separator: True**
But it still does not work, and I have no idea what's the use of this.

**Comment by TParcollet** [comment]
Hi,

The error that you are reporting does not seem to be related to SpeechBrain though as it is some OS decoding issue. Anyway, PreTraining is an extremely complex topic, simply because you can not come up with a universal solution. PreTraining is use-case based, meaning that there is an infinity of possibilities, and this is why our solution remains generic. 
We have a tutorial on the website: SpeechBrain Advanced 

If you are looking for more examples: simply search for:"# Load pretrained model if pretrained_separator is present in the yaml" in the search bar above, and you'll find plenty. Searching for: ".collect_files()" will also work. 

The best practice is to setup the PreTrainer in the yaml instead, as it helps with visibility when reading your code.

pretrained_separator should be an instance of Pretrained as written in the name. Meaning: provide a correctly setup Pretrained if you want to pretrain something.

For your question about only few layers, this is typically where coming with a universal solution is impossible. The clean way to do so is to initially break down your model into multiple nn.Module so it creates different .ckpt. If this is not the case, you must then either orchestrate your parameter transfer manually (as SB ckpt are nothing more than standard pytorch.save()) or load everything and re-initialise the layers of interest.
